tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/resource/N4JSResourceLoadStatesTest.xtend

```diff
@@ -0,0 +1,420 @@
+		FULLY_INITIALIZED_RECONCILED, FULLY_PROCESSED_RECONCILED
+	}
+	private static enum ParseResult {
+		NULL, AVAILABLE
+	}
+	private static enum AST {
+		NULL, PROXY, LAZY, AVAILABLE
+	}
+	private static enum Module {
+		NULL, STUBS, DEFERRED, AVAILABLE
+	}
+
+	@Data
+	static class StateInfo {
+		State state;
+		ParseResult parseResult;
+		AST ast;
+		Module module;
+		Boolean loaded;
+		Boolean fullyInitialized;
+		Boolean fullyProcessed;
+		Boolean reconciled;
+	}
+
+	static val Boolean INDETERMINATE = null;
+
+	static val loadStateInfos = #[
+		new StateInfo(State.CREATED,                      ParseResult.NULL,      AST.NULL,      Module.NULL,      false,          false, false, false ),
+		new StateInfo(State.CREATED_PRIME,                ParseResult.NULL,      AST.NULL,      Module.NULL,      false,          true,  false, false ),
+		new StateInfo(State.LOADED,                       ParseResult.AVAILABLE, AST.LAZY,      Module.NULL,      true,           false, false, false ),
+		new StateInfo(State.PRE_LINKED,                   ParseResult.AVAILABLE, AST.LAZY,      Module.STUBS,     true,           false, false, false ),
+		new StateInfo(State.FULLY_INITIALIZED,            ParseResult.AVAILABLE, AST.LAZY,      Module.DEFERRED,  true,           true,  false, false ),
+		new StateInfo(State.FULLY_PROCESSED,              ParseResult.AVAILABLE, AST.AVAILABLE, Module.AVAILABLE, true,           true,  true,  false ),
+		new StateInfo(State.LOADED_FROM_DESC,             ParseResult.NULL,      AST.PROXY,     Module.AVAILABLE, INDETERMINATE,  true,  true,  false ),
+		new StateInfo(State.LOADED_FROM_DESC_PRIME,       ParseResult.NULL,      AST.PROXY,     Module.DEFERRED,  INDETERMINATE,  true,  true,  false ),
+		new StateInfo(State.FULLY_INITIALIZED_RECONCILED, ParseResult.AVAILABLE, AST.LAZY,      Module.AVAILABLE, INDETERMINATE,  true,  false, true  ),
+		new StateInfo(State.FULLY_PROCESSED_RECONCILED,   ParseResult.AVAILABLE, AST.AVAILABLE, Module.AVAILABLE, INDETERMINATE,  true,  true,  true  )
+	];
+
+
+
+	// ###############################################################################################################
+	// TEST CASES
+
+	@Test
+	def void testStateCreated1() throws Exception {
+		val res = createFromFile(SAMPLE_FILE);
+		res.assertState(State.CREATED);
+	}
+
+	@Test
+	def void testStateCreated2() throws Exception {
+		val res = createFromFile(SAMPLE_FILE);
+		res.load(emptyMap);
+		res.unload();
+		res.assertState(State.CREATED);
+	}
+
+	@Test
+	def void testStateCreated3() throws Exception {
+		val res = createFromFile(SAMPLE_FILE);
+		res.load(emptyMap);
+		res.unloadAST();
+		res.assertState(State.CREATED);
+	}
+
+	@Test
+	def void testStateCreated4() throws Exception {
+		val res = createFromFile(SAMPLE_FILE);
+		res.load(emptyMap);
+		res.contents; // trigger installation of derived state (i.e. types builder)
+		// NOTE: for some reason, Xtext does not clear flag 'fullyInitialized' upon unload; that is done lazily when
```

Comment:
Probably not, as this would change Xtext's behavior and I suppose there was a reason for Xtext doing it the way it does. With this PR, we have at least documented this.

Reply:
But we overload the Xtext's `load` method and set the `fullyInitialized ourselves. What is the difference :-))